pkg/gardenlet/bootstrap/certificate/certificate_rotation_test.go

```diff
@@ -0,0 +1,342 @@
+					AddClient(keys.ForSeed(seed), k8sGardenClient)
+
+				_, err = getCurrentCertificate(context.TODO(), gardenClientConnection, fakeClientMap, seedName)
+				Expect(err).To(HaveOccurred())
+			})
+
+			It("should return an error - kubeconfig client cert is invalid", func() {
+				// mock existing secret with garden kubeconfig
+				updateKubeconfigSecret(k8sClient, gardenClientConnection, gardenKubeconfigWithInValidClientCert)
+
+				fakeClientMap := fakeclientmap.NewClientMap().
+					AddClient(keys.ForGarden(), k8sGardenClient).
+					AddClient(keys.ForSeed(seed), k8sGardenClient)
+
+				_, err := getCurrentCertificate(context.TODO(), gardenClientConnection, fakeClientMap, seedName)
+				Expect(err).To(HaveOccurred())
+			})
+		})
+
+		Describe("#waitForCertificateRotation", func() {
+			var (
+				fakeClientMap  clientmap.ClientMap
+				mockClient     *mockclient.MockClient
+				ctx            = context.TODO()
+				testKubeconfig string
+			)
+
+			BeforeEach(func() {
+				mockClient = mockclient.NewMockClient(ctrl)
+
+				validity := 10 * time.Second
+				cert := generateCertificate(validity)
+				testKubeconfig = fmt.Sprintf(baseKubeconfig, utils.EncodeBase64(cert.CertificatePEM), utils.EncodeBase64(cert.PrivateKeyPEM))
+
+				// mock first secret retrieval
+				mockClient.EXPECT().Get(ctx, kutil.Key(gardenClientConnection.KubeconfigSecret.Namespace, gardenClientConnection.KubeconfigSecret.Name), gomock.AssignableToTypeOf(&corev1.Secret{})).DoAndReturn(func(_ context.Context, _ client.ObjectKey, secret *corev1.Secret) error {
+					secret.ObjectMeta = metav1.ObjectMeta{
+						Name:      gardenClientConnection.KubeconfigSecret.Name,
+						Namespace: gardenClientConnection.KubeconfigSecret.Namespace,
+					}
+					secret.Data = map[string][]byte{kubernetes.KubeConfig: []byte(testKubeconfig)}
+					return nil
+				})
+
+				k8sGardenClient.EXPECT().DirectClient().Return(mockClient).AnyTimes()
+
+				fakeClientMap = fakeclientmap.NewClientMap().
+					AddClient(keys.ForGarden(), k8sGardenClient).
+					AddClient(keys.ForSeed(seed), k8sGardenClient)
+			})
+
+			It("should not return an error", func() {
+				// mock second secret retrieval - check the validity of the certificate again
+				// in this case the secret has not changed
+				mockClient.EXPECT().Get(ctx, kutil.Key(gardenClientConnection.KubeconfigSecret.Namespace, gardenClientConnection.KubeconfigSecret.Name), gomock.AssignableToTypeOf(&corev1.Secret{})).DoAndReturn(func(_ context.Context, _ client.ObjectKey, secret *corev1.Secret) error {
+					secret.ObjectMeta = metav1.ObjectMeta{
+						Name:      gardenClientConnection.KubeconfigSecret.Name,
+						Namespace: gardenClientConnection.KubeconfigSecret.Namespace,
+					}
+					secret.Data = map[string][]byte{kubernetes.KubeConfig: []byte(testKubeconfig)}
+					return nil
+				})
+
+				err := waitForCertificateRotation(context.TODO(), log, time.Now, gardenClientConnection, fakeClientMap, seedName)
```

Comment:
These test cases take pretty long:
```
------------------------------
• [SLOW TEST:7.975 seconds]
Certificates Tests that require a generated kubeconfig with a client certificate #waitForCertificateRotation should not return an error
/Users/d067603/go/src/github.com/gardener/gardener/pkg/gardenlet/bootstrap/certificate/certificate_rotation_test.go:276
------------------------------
• [SLOW TEST:7.547 seconds]
Certificates Tests that require a generated kubeconfig with a client certificate #waitForCertificateRotation should return an error - simulate changing the certificate while waiting for the certificate rotation deadline
/Users/d067603/go/src/github.com/gardener/gardener/pkg/gardenlet/bootstrap/certificate/certificate_rotation_test.go:292
------------------------------
```

Can you fake the `now` func to let it complete in the usual millisecond-range?

Reply:
The test itself is about: waiting for the  cert rotation to happen -  it is important for the test itself that it waits for a couple seconds (at the moment two tests at 7 seconds). We can decrease the timeout even more if needed.

Reply:
Decreased to one second so that the whole suite runs in < 2 seconds

```
Ran 8 of 8 Specs in 1.713 seconds
SUCCESS! -- 8 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestCertificates (1.71s)
```

Reply:
Yep, thanks!